### ISSUE TITLE
Suggestion: Listen on localhost by default

### ISSUE THREAD

**Issue by DavidBuchanan314** [memcached/memcached#425]
As I'm sure you're aware, the majority of memcached installations do not need to be exposed to the internet. And yet, there are tens of thousands of instances publicly accessible right now.

I would hazard a guess that a terrifyingly large percentage of those instances are also missing security patches.

Changing the default configuration to listen on localhost/127.0.0.1/::1 rather than `INADDR_ANY` will not magically make all these vulnerable servers go away, but I think it will significantly reduce the number of exposed instances in the future.

As a "case study", the Ubuntu Server repos already package memcached so that it listens on 127.0.0.1 by default. The Ubuntu builds also return e.g. `STAT version 1.4.25 Ubuntu`, so we can trivially scan the internet for these instances. According to my shodan search results, not even a single Ubuntu system is exposing memcached to the internet - ostensibly due to the safe(r) default configuration.

If there's a worry about people's installations breaking on upgrade, this seems like a good candidate for the `-o modern` option.

I'd be happy to make a relevant PR if this change would be welcome.

**Comment by dormando** [comment]
I'd probably have to give the project up to someone else to accomplish that. It'll cause years of breakage since people upgrade so slowly, and I don't have the time or the funding to cope with this. :(

Earlier this year I worked with vendors and just about everyone changed their default configurations; which was great since that only affects new installations or folks not customizing setups. If you change the default upstream it'll be rolling out behind the vendor changes anyway. So any extra benefit should be minimal.

However, you may have missed that we were able to disable UDP by default: https://github.com/memcached/memcached/commit/dbb7a8af90054bf4ef51f5814ef7ceb17d83d974 - since usage is so low, and that was the primary concern this year. I may also be able to lock the administrative commands to localhost by default without breaking much.

Thanks for your concern. Wish it weren't such a big deal, but we've done some seemingly minor changes to the defaults in the past and pretty much drove a huge section of the userbase away.

**Comment by DavidBuchanan314** [comment]
> It'll cause years of breakage since people upgrade so slowly

That may be true, but I'm hoping that the benefits outweigh the downsides here. Obviously I'm biased because I don't have to maintain/support anything, but this is my reasoning:

For simplicity, let's say there are 3 categories of memcached users:

1. Users who obtain their memcached builds/updates from a vendor, which defaults to listening on localhost (Great - they receive timely security updates, and they wouldn't be remotely exploitable by default anyway)

2. Users who obtain their memcached builds/updates from a vendor, which defaults to listening on `INADDR_ANY`. (OK - Will be vulnerable to RCE exploits if/when they appear, but will receive timely updates)

3. Users who downloaded a memcached binary/package from a random blog post etc., because their OS does not have a more "official" distribution channel. They do not receive security updates, and will inevitably sit waiting for someone to pwn them.

I'm assuming/hoping that category 1 is the large majority of memcached users, followed by 2 then 3. Category 1 will not be affected by my proposed changes, because it will already be their default configuration.

Category 2 *will* be affected - their vendors will have to adjust their packaged config, or users will have to adjust their local config, (but **only** if they really did depend on the default behaviour of listening on `INADDR_ANY`)

Although category 3 accounts for the smallest proportion of memcached users (hopefully), I believe it accounts for the majority of vulnerable instances on the public internet, right now.

In summary, I believe my proposed changes will have a negligible impact on the overall memcached userbase, but a very large positive impact on the security of "category 3" users, moving forwards.

RE: The UDP changes, I did see the commit, and that was why I was hoping these changes would be considered too. I'm not that concerned with the UDP reflection vulnerabilities anyway - my primary concern right now is that there are thousands of instances still vulnerable to older RCE exploits. My proposed changes won't fix this right away, but I hope it can reduce the next wave of mass exploitation in say, 10 years time.

**Comment by dormando** [comment]
It's a perfectly reasonable thing to do, that alienates users and causes a support nightmare. It doesn't actually solve the problem entirely either, as it's a network daemon and plenty of people will enable the networking and never secure it (how many mysql instances are open on the internet?). I have no idea how many of those people exist vs localhost installations.

While you're targeting (3), a small portion of users, every "real" user with networked pools of memcached's on secured networks will likely have to make an adjustment. If they get their updates solely through distro upgrades (as many do) that will be a surprise for them.

Nobody else works on this thing and I've no funding. If you (or anyone) were showing up as a dedicated volunteer to work on the project for a year that'd make it more likely to consider, but that's a pipe dream :)

**Comment by dormando** [comment]
what I really want to do is introduce even basic auth tokens to the client standard and eventually flip that as a default. This is one of the many nice things I'll likely never have.
